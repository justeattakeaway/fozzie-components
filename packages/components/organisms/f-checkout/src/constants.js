export const CHECKOUT_METHOD_COLLECTION = 'collection';
export const CHECKOUT_METHOD_DELIVERY = 'delivery';

export const VALID_CHECKOUT_METHODS = [CHECKOUT_METHOD_COLLECTION, CHECKOUT_METHOD_DELIVERY];

export const TENANT_MAP = {
    'en-GB': 'uk',
    'en-AU': 'au',
    'en-NZ': 'nz',
    'da-DK': 'dk',
    'es-ES': 'es',
    'en-IE': 'ie',
    'it-IT': 'it',
    'nb-NO': 'no'
};

export const VALIDATIONS = {
    address: 'addressValidations',
    guest: 'guestValidations'
};
