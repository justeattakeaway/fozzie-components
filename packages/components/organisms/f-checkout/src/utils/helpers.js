/**
 * Gets a specific cookie based on the cookie name you provide it.
 *
 * TODO: Use universal cookie helper when consuming apps are ready.
 *
 * @param name
 * @returns {*}
 */
const getCookie = name => {
    const nameEQ = `${name}=`;
    const ca = document.cookie.split(';');
    for (let i = 0; i < ca.length; i++) {
        let c = ca[i];
        while (c.charAt(0) === ' ') c = c.substring(1, c.length);
        if (c.indexOf(nameEQ) === 0) return c.substring(nameEQ.length, c.length);
    }
    return null;
};

export {
    // eslint-disable-next-line import/prefer-default-export
    getCookie
};
