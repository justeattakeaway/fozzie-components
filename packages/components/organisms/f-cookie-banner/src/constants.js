export const CONSENT_COOKIE_NAME = 'je-cookieConsent';
export const LEGACY_CONSENT_COOKIE_NAME = 'je-banner_cookie';
