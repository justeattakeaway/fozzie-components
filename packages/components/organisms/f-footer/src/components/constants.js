export const confianzaUrl = 'https://www.confianzaonline.es/empresas/justeatspain.htm';
export const confianzaScreenReaderText = 'Accede a Confianza Online y comprueba nuestro sello de calidad en Internet';
