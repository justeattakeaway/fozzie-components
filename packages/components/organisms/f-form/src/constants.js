export const FORM_EVENTS = {
    fieldUpdated: 'updated',
    submitting: 'form-submitting'
};

export const PROP_VALIDATION_MESSAGES = {
    requiredFormFields: '`formFields` Not found in `formData` prop, please add an array of `formFields`',
    invalidFormFields: '`formFields` are invalid, please add an array of formFields',
    requiredNameProperty: 'Each field in `formFields` should contain "name" property',
    requiredTranslationsProperty: 'Each field in `formFields` should contain "translations" property',
    requiredLabelProperty: 'Each field in `formFields` should contain "translations.label" property'
};

export const DEFAULT_BUTTON_TEXT = 'Submit';
