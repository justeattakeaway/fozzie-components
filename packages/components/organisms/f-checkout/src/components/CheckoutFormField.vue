<template>
    <form-field
        :value="value"
        :name="translations.name"
        :input-type="inputType"
        :label-text="translations.label"
        :has-error="hasError"
        :maxlength="maxLength"
        v-bind="isGrouped && groupedProps"
        :aria-invalid="hasError"
        :aria-describedby="hasValidationMessages && translations.errorName"
        @blur="shouldValidateOnBlur && formFieldBlur()"
        @input="updateUserDetails({ fieldType, fieldName, value: $event })">
        <template #error>
            <error-message
                v-if="hasErrorType('required') && isEmpty"
                :id="translations.errorName"
                data-js-error-message
                :class="[{ [$style['c-checkoutFormField-error--grouped']]: isGrouped }]"
                :data-test-id="translations.emptyError">
                {{ translations.validationMessages.required }}
            </error-message>
            <error-message
                v-else-if="!isValid"
                :id="translations.errorName"
                data-js-error-message
                :data-test-id="translations.invalidError">
                {{ translations.validationMessages.invalid }}
            </error-message>
        </template>
    </form-field>
</template>

<script>
import ErrorMessage from '@justeat/f-error-message';
import '@justeat/f-error-message/dist/f-error-message.css';
import FormField from '@justeat/f-form-field';
import '@justeat/f-form-field/dist/f-form-field.css';
import { mapState, mapActions } from 'vuex';
import checkoutValidationsMixin from '../mixins/validations.mixin';
import { VALIDATIONS, VUEX_CHECKOUT_MODULE } from '../constants';

export default {
    components: { FormField, ErrorMessage },
    mixins: [checkoutValidationsMixin],
    props: {
        fieldName: {
            type: String,
            required: true
        },

        fieldType: {
            type: String,
            required: true
        },

        inputType: {
            type: String,
            default: 'text'
        },

        maxLength: {
            type: String,
            default: '100'
        },

        shouldValidateOnBlur: {
            type: Boolean,
            default: false
        },

        isGrouped: {
            type: Boolean,
            defalut: false
        }
    },

    /**
     * Provide/Inject allows component to inherit `Checkout`
     * validator scope, `$v`.
    */
    inject: ['$v'],

    computed: {
        ...mapState(VUEX_CHECKOUT_MODULE, [
            'customer',
            'address'
        ]),

        value () {
            return this[this.fieldType][this.fieldName];
        },

        translations () {
            return {
                ...this.$t(`formFields.${this.fieldType}.${this.fieldName}`),
                name: `${this.kebabCase}`,
                errorName: `${this.kebabCase}-error`,
                emptyError: `error-${this.kebabCase}-empty`,
                invalidError: `error-${this.kebabCase}-invalid`
            };
        },

        isEmpty () {
            return this.isFieldEmpty(VALIDATIONS[this.fieldType], this.fieldName);
        },

        isValid () {
            if (this.hasErrorType('invalid')) {
                const validations = this.$v[VALIDATIONS[this.fieldType]][this.fieldName];

                return (!validations.$dirty || validations[this.fieldName]) && !this.isEmpty;
            }
            return true;
        },

        hasError () {
            return this.hasValidationMessages && (this.isEmpty || !this.isValid);
        },

        hasValidationMessages () {
            return !!this.translations.validationMessages;
        },

        kebabCase () {
            return this.fieldName.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, '$1-$2').toLowerCase();
        },

        groupedProps () {
            return {
                placeholder: this.translations.label,
                isGrouped: true,
                shouldShowLabelText: false
            };
        }
    },

    methods: {
        ...mapActions(VUEX_CHECKOUT_MODULE, [
            'updateUserDetails'
        ]),

        formFieldBlur () {
            const fieldValidation = this.$v[VALIDATIONS[this.fieldType]][this.fieldName];

            if (fieldValidation) {
                fieldValidation.$touch();
            }
        },

        hasErrorType (errorType) {
            return this.hasValidationMessages && !!this.translations.validationMessages[errorType];
        }
    }
};
</script>

<style lang="scss" module>
.c-checkoutFormField-error--grouped {
    margin-bottom: spacing(x2);
}
</style>
