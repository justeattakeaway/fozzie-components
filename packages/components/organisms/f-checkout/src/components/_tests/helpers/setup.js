import Vuex from 'vuex';
import { CHECKOUT_METHOD_DELIVERY } from '../../../constants';
import tenantConfigs from '../../../tenants';

const fulfilmentTimes = [
    {
        from: '2020-01-01T01:00:00.000Z',
        to: '2020-01-01T01:00:00.000Z'
    },
    {
        from: '2020-01-01T01:15:00.000Z',
        to: '2020-01-01T01:15:00.000Z'
    }
];

const defaultCheckoutState = {
    id: '',
    serviceType: CHECKOUT_METHOD_DELIVERY,
    restaurantId: '',
    basket: {
        id: '',
        total: 0
    },
    customer: {
        firstName: 'John',
        lastName: 'Smith',
        email: 'john@test.com',
        mobileNumber: '+447111111111'
    },
    orderId: 'ORDER111',
    time: {
        from: '',
        to: ''
    },
    address: {
        line1: '1 Bristol Road',
        line2: 'Flat 1',
        city: 'Bristol',
        postcode: 'BS1 1AA'
    },
    availableFulfilment: {
        times: fulfilmentTimes,
        isAsapAvailable: true
    },
    isFulfillable: true,
    notices: [],
    messages: [],
    authToken: '',
    isLoggedIn: false,
    userNote: 'No ketchup, please.',
    geolocation: null
};

const defaultAnalyticsState = {
    autofill: [],
    changedFields: []
};

const defaultCheckoutActions = {
    getCheckout: jest.fn(),
    updateCheckout: jest.fn(),
    getAvailableFulfilment: jest.fn(),
    setAuthToken: jest.fn(),
    createGuestUser: jest.fn(),
    updateAddressDetails: jest.fn(),
    updateCustomerDetails: jest.fn(),
    updateFulfilmentTime: jest.fn(),
    getGeoLocation: jest.fn(),
    getBasket: jest.fn(),
    getAddress: jest.fn(),
    placeOrder: jest.fn()
};

const defaultAnalyticsActions = {
    updateAutofill: jest.fn(),
    updateChangedField: jest.fn(),
    trackInitialLoad: jest.fn(),
    trackFormInteraction: jest.fn()
};

const i18n = {
    locale: 'en-GB',
    messages: {
        'en-GB': tenantConfigs['en-GB'].messages
    },
    dateTimeFormats: {
        'en-GB': tenantConfigs['en-GB'].dateTimeFormats
    }
};

const createStore = (
    checkoutState = defaultCheckoutState,
    checkoutActions = defaultCheckoutActions,
    analyticsState = defaultAnalyticsState,
    analyticsActions = defaultAnalyticsActions
) => new Vuex.Store({
    modules: {
        checkout: {
            namespaced: true,
            state: checkoutState,
            actions: checkoutActions
        },
        analytics: {
            namespaced: true,
            state: analyticsState,
            actions: analyticsActions
        },
        hasModule: jest.fn(() => true)
    }
});

const $logger = {
    logInfo: jest.fn(),
    logWarn: jest.fn(),
    logError: jest.fn()
};

export {
    fulfilmentTimes,
    defaultCheckoutState,
    defaultCheckoutActions,
    defaultAnalyticsState,
    i18n,
    createStore,
    $logger
};
