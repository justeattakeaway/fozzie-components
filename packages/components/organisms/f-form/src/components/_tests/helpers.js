import { createLocalVue } from '@vue/test-utils';
import { VueI18n } from '@justeat/f-globalisation';
import tenantConfigs from '../../tenants';
import { fieldWithRequiredValidations } from '../../../stories/helpers';

const localVue = createLocalVue();
localVue.use(VueI18n);

const i18n = {
    locale: 'en-GB',
    messages: {
        'en-GB': tenantConfigs['en-GB'].messages
    }
};

const $v = {
    formFields: {
        mobileNumber: {
            $dirty: false,
            mobileNumber: false
        },
        postCode: {
            $dirty: false,
            postCode: false
        },
        firstName: {
            $dirty: false,
            $touch: jest.fn()
        },
        lastName: {
            $dirty: false
        },
        email: {
            $dirty: false,
            email: false
        }
    },
    dirty: false,
    $touch: jest.fn()
};

const FormFieldMock = {
    props: {
        name: 'first-name',
        'label-text': 'First Name',
        'input-type': 'text'
    },
    translations: fieldWithRequiredValidations.translations,
    validationMessages: fieldWithRequiredValidations.translations.validationMessages,
    value: fieldWithRequiredValidations.value
};

const FormFieldErrorMock = {
    props: {
        maxLength: null,
        name: 'firstName',
        'label-text': 'First Name',
        'input-type': 'text',
        'max-length': null,
        'aria-invalid': true,
        'has-error': true,
        'aria-describedby': 'error-postcode-required'
    },
    errorMessage: {
        props: {
            'data-test-id': 'error-postcode-required',
            id: 'error-postcode-required'
        },
        text: fieldWithRequiredValidations.translations.validationMessages.required
    },
    translations: fieldWithRequiredValidations.translations,
    validationMessages: fieldWithRequiredValidations.translations.validationMessages,
    value: fieldWithRequiredValidations.value
};

export {
    $v,
    localVue,
    i18n,
    FormFieldMock,
    FormFieldErrorMock
};
