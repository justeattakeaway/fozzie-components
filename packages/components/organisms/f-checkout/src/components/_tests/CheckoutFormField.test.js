import Vuex from 'vuex';
import { VueI18n } from '@justeat/f-globalisation';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import CheckoutFormField from '../CheckoutFormField.vue';
import { i18n, createStore } from './helpers/setup';
import checkoutValidationsMixin from '../../mixins/validations.mixin';
import { VALIDATIONS } from '../../constants';

const localVue = createLocalVue();

localVue.use(VueI18n);
localVue.use(Vuex);

const $v = {
    [VALIDATIONS.guest]: {
        firstName: {
            $dirty: false
        },
        lastName: {
            $dirty: false
        },
        email: {
            $dirty: false,
            email: false,
            $touch: jest.fn()
        }
    }
};

const $t = jest.fn();

describe('CheckoutFormField', () => {
    const propsData = {
        fieldName: 'firstName'
    };

    it('should be defined', () => {
        // Arrange
        const wrapper = shallowMount(CheckoutFormField, {
            i18n,
            store: createStore(),
            localVue,
            propsData,
            provide: () => ({
                $v
            })
        });

        // Assert
        expect(wrapper.exists()).toBe(true);
    });

    describe('computed ::', () => {
        describe('isEmpty ::', () => {
            let isFieldEmptySpy;

            beforeEach(() => {
                // Arrange
                isFieldEmptySpy = jest.spyOn(checkoutValidationsMixin.methods, 'isFieldEmpty');
            });

            afterEach(() => {
                jest.clearAllMocks();
            });

            it.each([
                ['firstName'],
                ['lastName'],
                ['email']
            ])('should call `isFieldEmpty` with %s and %s', fieldName => {
                // Act
                shallowMount(CheckoutFormField, {
                    i18n,
                    store: createStore(),
                    localVue,
                    propsData: {
                        fieldName
                    },
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(isFieldEmptySpy).toHaveBeenCalledWith(VALIDATIONS.guest, fieldName);
            });
        });

        describe('isValid ::', () => {
            afterEach(() => {
                jest.restoreAllMocks();
            });

            describe('when `hasErrorType` returns `false`', () => {
                let wrapper;

                beforeEach(() => {
                    // Arrange & Act
                    wrapper = shallowMount(CheckoutFormField, {
                        i18n,
                        store: createStore(),
                        localVue,
                        propsData,
                        provide: () => ({
                            $v
                        })
                    });

                    jest.spyOn(wrapper.vm, 'hasErrorType').mockReturnValue(false);
                });

                afterEach(() => {
                    jest.restoreAllMocks();
                });

                it('should return true', () => {
                    // Assert
                    expect(wrapper.vm.isValid).toEqual(true);
                });
            });

            describe('when `hasErrorType` is `true`', () => {
                describe('AND `isEmpty` is `true`', () => {
                    beforeEach(() => {
                        // Arrange
                        jest.spyOn(CheckoutFormField.methods, 'hasErrorType').mockImplementation(() => true);
                    });

                    afterEach(() => {
                        jest.restoreAllMocks();
                    });

                    it('should return `false`', () => {
                        // Act
                        const wrapper = shallowMount(CheckoutFormField, {
                            i18n,
                            store: createStore(),
                            localVue,
                            propsData,
                            provide: () => ({
                                $v
                            }),
                            computed: {
                                isEmpty () {
                                    return true;
                                }
                            }
                        });

                        // Assert
                        expect(wrapper.vm.isValid).toEqual(false);
                    });
                });

                describe('AND `isEmpty` is `false`', () => {
                    describe.each([
                        [true, true, true],
                        [false, true, true],
                        [false, false, true],
                        [true, false, false]
                    ])('AND `$dirty` is `%s`, field validation is `%s`', ($dirty, isValid, expected) => {
                        let wrapper;

                        beforeEach(() => {
                            // Arrange & Act
                            $v[VALIDATIONS.guest].email = {
                                $touch: jest.fn(),
                                $dirty,
                                email: isValid
                            };

                            wrapper = shallowMount(CheckoutFormField, {
                                i18n,
                                store: createStore(),
                                localVue,
                                propsData: {
                                    fieldName: 'email'
                                },
                                provide: () => ({
                                    $v
                                }),
                                computed: {
                                    isEmpty () {
                                        return false;
                                    }
                                }
                            });

                            jest.spyOn(wrapper.vm, 'hasErrorType').mockReturnValue(true);
                        });

                        afterEach(() => {
                            jest.restoreAllMocks();
                        });

                        it('should return `%s`', () => {
                            // Assert
                            expect(wrapper.vm.isValid).toEqual(expected);
                        });
                    });
                });
            });
        });

        describe('kebabCase ::', () => {
            it.each([
                ['firstName', 'first-name'],
                ['lastName', 'last-name'],
                ['email', 'email']
            ])('should convert `fieldName` value to kebab case', (fieldName, expected) => {
                // Arrange & Act
                const wrapper = shallowMount(CheckoutFormField, {
                    i18n,
                    store: createStore(),
                    localVue,
                    propsData: {
                        fieldName
                    },
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(wrapper.vm.kebabCase).toEqual(expected);
            });
        });
    });

    describe('methods ::', () => {
        afterEach(() => {
            jest.clearAllMocks();
        });

        describe('formFieldBlur ::', () => {
            describe('when `shouldValidateOnBlur` is `true` AND field has validations', () => {
                let touchSpy;

                beforeEach(() => {
                    touchSpy = jest.spyOn($v[VALIDATIONS.guest].email, '$touch');
                });

                afterEach(() => {
                    jest.clearAllMocks();
                });

                it('should call `touch` for validation', () => {
                    // Arrange
                    const fieldName = 'email';

                    const wrapper = shallowMount(CheckoutFormField, {
                        i18n,
                        store: createStore(),
                        localVue,
                        propsData: {
                            ...propsData,
                            fieldName,
                            shouldValidateOnBlur: true
                        },
                        provide: () => ({
                            $v
                        })
                    });

                    // Act
                    wrapper.vm.formFieldBlur(fieldName);

                    // Assert
                    expect(touchSpy).toHaveBeenCalled();
                });
            });
        });

        describe('hasErrorType ::', () => {
            const invalidErrorType = 'invalid';
            const requiredErrorType = 'required';

            const fieldWithRequiredError = 'firstName';
            const fieldWithInvalidError = 'lastName';

            const translations = {
                formFields: {
                    guest: {
                        [fieldWithRequiredError]: {
                            label: 'First Name',
                            validationMessages: {
                                required: 'Has required error'
                            }
                        },
                        [fieldWithInvalidError]: {
                            label: 'Last Name',
                            validationMessages: {
                                invalid: 'Has invalid error'
                            }
                        }
                    }
                }
            };

            afterEach(() => {
                jest.clearAllMocks();
            });

            it.each([
                [true, fieldWithRequiredError, requiredErrorType],
                [false, fieldWithRequiredError, invalidErrorType],
                [false, fieldWithInvalidError, requiredErrorType],
                [true, fieldWithInvalidError, invalidErrorType]
            ])('should return `%s` when `hasErrorType` `fieldName` is %s and `errorType` is %s', (expected, fieldName, errorType) => {
                // Arrange
                $t.mockReturnValue(translations.formFields.guest[fieldName]);

                const wrapper = shallowMount(CheckoutFormField, {
                    i18n,
                    store: createStore(),
                    localVue,
                    propsData: {
                        fieldName
                    },
                    provide: () => ({
                        $v
                    }),
                    mocks: {
                        $t
                    }
                });

                // Act
                const received = wrapper.vm.hasErrorType(errorType);

                // Assert
                expect(received).toEqual(expected);
            });
        });
    });
});
