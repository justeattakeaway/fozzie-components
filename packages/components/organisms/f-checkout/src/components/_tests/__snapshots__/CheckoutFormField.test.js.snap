// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutFormField computed :: groupedProps :: should return props for grouped form fields 1`] = `
Object {
  "isGrouped": true,
  "placeholder": undefined,
  "shouldShowLabelText": false,
}
`;

exports[`CheckoutFormField computed :: translations :: should return translations from with id names 1`] = `
Object {
  "emptyError": "error-first-name-empty",
  "errorName": "first-name-error",
  "invalidError": "error-first-name-invalid",
  "label": "First name",
  "name": "first-name",
  "validationMessages": Object {
    "required": "Please enter your first name",
  },
}
`;
